rust_src/src/multibyte.rs

```diff
@@ -110,6 +110,34 @@ impl LispStringRef {
     pub fn byte_at(&self, index: ptrdiff_t) -> u8 {
         unsafe { *self.const_data_ptr().offset(index) }
     }
+
+    /// (Somewhat) Replaces STRING_SET_UNIBYTE in C. That macro has the behavior that
+    /// if your string has size 0, it will replace your string variable with
+    /// 'empty_unibyte_string'. This function does not have that side-effect.
+    #[inline]
+    pub fn set_unibyte(mut self) {
+        if self.size != 0 {
+            self.size_byte = -1;
+        }
+    }
+
+    /// This function does not allocate. It will not change the size of the data allocation.
+    /// It will only set the 'size' variable of the string, if it is safe to do so.
+    /// Replaces STRING_SET_CHARS from C.
+    pub fn set_num_chars(mut self, newsize: isize) {
+        assert!(if self.is_multibyte() {
```

Comment:
Is assert really what we want long term? The C code uses `eassert` which can be disabled. In Rust `assert!` is always on.

Reply:
So this was an intentional decision by me that I should have clarified in the PR description. 

Basically I was very torn if this function should be marked 'unsafe'. At first glance this function seemed like a footgun to me. It's name/semantics seem to imply it's doing an potential allocation, when it's actually it's just assigning to a variable, AND we only want that assignment to be possible in a very specific situation (what is checked for in the eassert). 

I did some basic profiling of "clear_string", in emacs, release remacs with this being a debug_assert, and remacs with this being an assert, and all 3 were almost identical. With that information, I felt that the best course of action was to make this an "assert" and mark the function as safe. 

If we would like to switch this to be a "debug_assert" to keep in line with the C code, I would like to insist that this function be marked unsafe.

Reply:
So let's kill the function. It is called in exactly 3 places in the emacs source. I like marking this as "unsafe" until then.

Reply:
Sounds good to me. 👍

Reply:
One place is dired. There is an open PR to help finish.
One place is this PR. Natch.
One place is lread.c